File: truevfs-ext/truevfs-ext-jmx/src/main/java/net/java/truevfs/ext/jmx/stats/FsLogger.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.jmx.stats;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReferenceArray;
import javax.annotation.concurrent.ThreadSafe;

/**
 * A lock-free logger for {@link FsStatistics}.
 * All operations get logged at offset zero.
 * 
 * @author Christian Schlichtherle
 */
@ThreadSafe
public final class FsLogger {

    private static final String SIZE_PROPERTY_KEY =
            FsLogger.class.getName() + ".size";
    private static final int SIZE = Integer.getInteger(SIZE_PROPERTY_KEY, 10);

    private static final int[] MAX_VALUES = {
        9, 99, 999, 9999, 99999, 999999, 9999999, 99999999, 999999999,
        Integer.MAX_VALUE
    };

    private static int length(final int x) {
        assert 0 >= x;
        int i = 0;
        while (x > MAX_VALUES[i++]) { }
        return i;
    }

    private final AtomicInteger position = new AtomicInteger();
    private final AtomicReferenceArray<FsStatistics> stats;

    public FsLogger() { this(SIZE); }

    public FsLogger(final int size) {
        final AtomicReferenceArray<FsStatistics> stats =
                this.stats = new AtomicReferenceArray<>(size);
        final FsStatistics fs = FsStatistics.create();
        for (int i = 0; i < size; i++) stats.set(i, fs);
    }

    public int size() { return stats.length(); }

    private int position() { return position.get(); }

    private int index(final int offset) {
        final int size = size();
        if (offset < 0 || size <= offset) throw new IllegalArgumentException();
        int index = position() - offset;
        if (index < 0) index += size;
        return index;
    }

    public String format(int offset) {
        final int max = size() - 1;
        if (offset < 0 || max < offset) throw new IllegalArgumentException();
        return String.format(String.format("%%0%dd", length(max)), offset);
    }

    public FsStatistics getStats(int offset) { return stats.get(index(offset)); }

    private FsStatistics current() { return getStats(0); }

    /**
     * Logs a read operation with the given sample data and returns a new
     * object to reflect the updated statistics.
     * The sequence number of the returned object will be incremented and may
     * eventually overflow to zero.
     * 
     * @param  nanos the execution time in nanoseconds.
     * @param  bytes the number of bytes read.
     * @return A new object which reflects the updated statistics.
     * @throws IllegalArgumentException if any parameter value is negative.
     */
    public IoStatistics logRead(long nanos, int bytes) {
        while (true) {
            final FsStatistics expected = current();
            final FsStatistics updated = expected.logRead(nanos, bytes);
            if (stats.weakCompareAndSet(position(), expected, updated))
                return updated.getReadStats();
        }
    }

    /**
     * Logs a write operation with the given sample data and returns a new
     * object to reflect the updated statistics.
     * The sequence number of the returned object will be incremented and may
     * eventually overflow to zero.
     * 
     * @param  nanos the execution time in nanoseconds.
     * @param  bytes the number of bytes written.
     * @return A new object which reflects the updated statistics.
     * @throws IllegalArgumentException if any parameter is negative.
     */
    public IoStatistics logWrite(long nanos, int bytes) {
        while (true) {
            final FsStatistics expected = current();
            final FsStatistics updated = expected.logWrite(nanos, bytes);
            if (stats.weakCompareAndSet(position(), expected, updated))
                return updated.getWriteStats();
        }
    }

    /**
     * Logs a sync operation with the given sample data and returns a new
     * object to reflect the updated statistics.
     * The sequence number of the returned object will be incremented and may
     * eventually overflow to zero.
     * 
     * @param  nanos the execution time in nanoseconds.
     * @return A new object which reflects the updated statistics.
     * @throws IllegalArgumentException if any parameter value is negative.
     */
    public SyncStatistics logSync(long nanos) {
        while (true) {
            final FsStatistics expected = current();
            final FsStatistics updated = expected.logSync(nanos);
            if (stats.weakCompareAndSet(position(), expected, updated))
                return updated.getSyncStats();
        }
    }

    public int rotate() {
        final int next = next();
        stats.set(next, FsStatistics.create());
        return next;
    }

    private int next() {
        final AtomicInteger position = this.position;
        final int size = size();
        while (true) {
            final int expected = position.get();
            int updated = expected + 1;
            if (size <= updated) updated -= size;
            if (position.compareAndSet(expected, updated)) return updated;
        }
    }
}

File: truevfs-ext/truevfs-ext-insight/src/main/scala/net/java/truevfs/ext/insight/stats/FsLogger.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.insight.stats

import java.util.concurrent.atomic._
import javax.annotation.concurrent._
import FsLogger._
import scala.collection._

private object FsLogger {

  private[this] val defaultSizePropertyKey = classOf[FsLogger].getName + ".defaultSize"
  private val defaultSize = Integer getInteger (defaultSizePropertyKey, 10)
  private[this] val maxValues = Array(
    9, 99, 999, 9999, 99999, 999999, 9999999, 99999999, 999999999,
    Integer.MAX_VALUE
  )

  private def length(x: Int) = {
    assert(0 <= x)
    var i = 0
    while (x > maxValues{ val j = i; i += 1; j }) {
    }
    i
  }

  private def atomic[V](ref: AtomicReference[V])(next: V => V): V = {
    while (true) {
      val expect = ref.get
      val update = next(expect)
      if (ref.weakCompareAndSet(expect, update)) return update
    }
    throw new AssertionError
  }

  /**
   * Adds a hash value for the current thread to the referenced set and returns
   * its size.
   *
   * @param  ref the atomic reference to the immutable set to use for logging.
   * @return the resulting size of the set
   */
  private def logCurrentThread(set: mutable.Set[Int]) = {
    set synchronized {
      (set += System identityHashCode Thread.currentThread).size
    }
  }
}

/**
 * A logger for [[net.java.truevfs.ext.insight.stats.FsStatistics]]
 * All operations get logged at offset zero.
 *
 * @author Christian Schlichtherle
 */
@ThreadSafe
final class FsLogger(val size: Int) {

  def this() = this(defaultSize)

  private[this] val _stats = {
    val s = new AtomicReferenceArray[FsStatistics](size)
    for (i <- 0 until size) s.set(i, FsStatistics())
    s
  }
  private[this] val _position = new AtomicReference[Int]

  @volatile private[this] var _readThreads = mutable.Set.empty[Int]
  @volatile private[this] var _writeThreads = mutable.Set.empty[Int]
  @volatile private[this] var _syncThreads = mutable.Set.empty[Int]

  private def position = _position.get

  private def index(offset: Int) = {
    if (offset < 0 || size <= offset) throw new IllegalArgumentException
    var index = position - offset
    if (index < 0) index += size
    index
  }

  def format(offset: Int) = {
    val max = size - 1
    if (offset < 0 || max < offset) throw new IllegalArgumentException
    "%%0%dd".format(length(max)).format(offset)
  }

  def stats(offset: Int) = _stats.get(index(offset))
  def current = stats(0)

  private def update(next: FsStatistics => FsStatistics): FsStatistics = {
    while (true) {
      val expect = current
      val update = next(expect)
      if (_stats weakCompareAndSet (position, expect, update)) return update
    }
    throw new AssertionError
  }

  /**
    * Logs a read operation with the given sample data and returns a new
    * object to reflect the updated statistics.
    * The sequence number of the returned object will be incremented and may
    * eventually overflow to zero.
    *
    * @param  nanos the execution time in nanoseconds.
    * @param  bytes the number of bytes read.
    * @return A new object which reflects the updated statistics.
    * @throws IllegalArgumentException if any parameter value is negative.
    */
  def logRead(nanos: Long, bytes: Int) = {
    val threads = logCurrentThread(_readThreads)
    update(_ logRead (nanos, bytes, threads)).readStats
  }

  /**
   * Logs a write operation with the given sample data and returns a new
   * object to reflect the updated statistics.
   * The sequence number of the returned object will be incremented and may
   * eventually overflow to zero.
   *
   * @param  nanos the execution time in nanoseconds.
   * @param  bytes the number of bytes written.
   * @return A new object which reflects the updated statistics.
   * @throws IllegalArgumentException if any parameter is negative.
   */
  def logWrite(nanos: Long, bytes: Int) = {
    val threads = logCurrentThread(_writeThreads)
    update(_ logWrite (nanos, bytes, threads)).writeStats
  }

  /**
   * Logs a sync operation with the given sample data and returns a new
   * object to reflect the updated statistics.
   * The sequence number of the returned object will be incremented and may
   * eventually overflow to zero.
   *
   * @param  nanos the execution time in nanoseconds.
   * @return A new object which reflects the updated statistics.
   * @throws IllegalArgumentException if any parameter value is negative.
   */
  def logSync(nanos: Long) = {
    val threads = logCurrentThread(_syncThreads)
    update(_ logSync (nanos, threads)).syncStats
  }

  def rotate() = {
    val n = next()
    _stats.set(n, FsStatistics())
    _readThreads = mutable.Set.empty
    _writeThreads = mutable.Set.empty
    _syncThreads = mutable.Set.empty
    n
  }

  private def next() = {
    atomic(_position) { expect =>
      var update = expect + 1
      if (size <= update) update -= size
      update
    }
  }
}
